File: code/ex752.java

// BFS

class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> lock = new HashSet<>();
        for (String deadend : deadends)    lock.add(deadend);
        
        Queue<String> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        queue.offer("0000");    if (target.equals("0000"))    return 0;     
        if (lock.contains("0000"))    return -1;
        visited.add("0000");
        
        int step = 0;
        while (!queue.isEmpty()) {
            for (int z = queue.size(); z > 0; --z) {
                String out = queue.poll();
                
                // Try to turn each wheel two directions
                for (int i = 0; i < 4; i++) {
                    // Turn each wheel forward
                    StringBuilder sb = new StringBuilder(out);
                    sb.setCharAt(i, (char)((sb.charAt(i) - '0' + 1) % 10 + '0'));
                    String newStr = sb.toString();
                    if (newStr.equals(target))    return step + 1;
                    if (!visited.contains(newStr) && !lock.contains(newStr)) {
                        queue.offer(newStr);
                        visited.add(newStr);
                    }
                    
                    // Turn each wheel backward
                    sb = new StringBuilder(out);
                    sb.setCharAt(i, (char)((sb.charAt(i) - '0' + 9) % 10 + '0'));
                    newStr = sb.toString();
                    if (newStr.equals(target))    return step + 1;
                    if (!visited.contains(newStr) && !lock.contains(newStr)) {
                        queue.offer(newStr);
                        visited.add(newStr);
                    }
                }
            }
            step++;
        }
        
        return -1;
    }
}

File: code/ex752.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> dict(deadends.begin(), deadends.end());
        
        if (dict.count("0000") > 0)    return -1;
        queue<string> que;
        que.push("0000");
        dict.insert("0000");
        
        int step = 0;
        while (!que.empty()) {
            int size = que.size();
            for (int z = size; z > 0; z--) {
                string out = que.front();    que.pop();
                if (out == target)    return step;
                for (int i = 0; i < 4; i++) {
                    string bkp = out;
                    // Add 1
                    if (bkp[i] == '9')    bkp[i] = '0';
                    else    bkp[i]++;
                    if (dict.count(bkp) == 0) {
                        que.push(bkp);
                        dict.insert(bkp);
                    }
                    
                    // Minus 1
                    bkp = out;
                    if (bkp[i] == '0')    bkp[i] = '9';
                    else    bkp[i]--;
                    if (dict.count(bkp) == 0) {
                        que.push(bkp);
                        dict.insert(bkp);
                    }
                }
            }
            step++;
        }
        
        return -1;
    }
};